### PR TITLE
Add MONAI quickstart example

### PR DESCRIPTION
**Pull request by charlesbvll**
### Description

[MONAI](https://docs.monai.io/en/latest/index.html) (Medical Open Network for AI) is a PyTorch-based, open-source framework for deep learning in healthcare imaging, part of the PyTorch Ecosystem and we don't yet have an example for it.

### Related issues/PRs

N/A

### Explanation

Add a quickstart example using Flower and MONAI.

### Checklist

- [x] Implement proposed change
- [x] Update the changelog entry below
- [x] Make CI checks pass
- [x] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### Changelog entry

[MONAI](https://docs.monai.io/en/latest/index.html) (Medical Open Network for AI) is a PyTorch-based, open-source framework for deep learning in healthcare imaging, part of the PyTorch Ecosystem. You can find the example at https://github.com/adap/flower/tree/main/examples/quickstart-monai.

### Any other comments?

N/A

### CODE REVIEW

File: examples/quickstart-monai/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+flwr>=1.0, <2.0
+torch==1.13.1
+tqdm==4.65.0
+scikit-learn==1.3.1
+monai[gdown,nibabel,tqdm,itk]==1.3.0
+numpy==1.24.4
```
Comment (jafermarq) [inline examples/quickstart-monai/requirements.txt:6]:
```suggestion
numpy==1.24.4
pillow==10.2.0
```

File: examples/quickstart-monai/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+authors = ["The Flower Authors <hello@flower.dev>"]
+
+[tool.poetry.dependencies]
+python = ">=3.8,<3.11"
+flwr = ">=1.0,<2.0"
+torch = "1.13.1"
+tqdm = "4.65.0"
+scikit-learn = "1.3.1"
+monai = { version = "1.3.0", extras=["gdown", "nibabel", "tqdm", "itk"] }
+numpy = "1.21.3"
```
Comment (jafermarq) [inline examples/quickstart-monai/pyproject.toml:18]:
```suggestion
numpy = "1.24.4"
pillow = "10.2.0"
```

File: examples/quickstart-monai/README.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+## Run Federated Learning with MONAI and Flower
+
+Afterwards you are ready to start the Flower server as well as the clients. You can simply start the server in a terminal as follows:
+
+```shell
+python3 server.py
+```
+
+Now you are ready to start the Flower clients which will participate in the learning. To do so simply open two more terminal windows and run the following commands.
```
Comment (jafermarq) [inline examples/quickstart-monai/README.md:71]:
```suggestion
Now you are ready to start the Flower clients which will participate in the learning. To do so simply open two more terminal windows and run the following commands.  Clients will train a [DenseNet121](https://docs.monai.io/en/stable/networks.html#densenet121) from MONAI. If a GPU is present in your system, clients will use it.
```

File: examples/quickstart-monai/pyproject.toml (hunk 2 above)
Comment (jafermarq) [reply examples/quickstart-monai/pyproject.toml:18]: adding pillow. Raising `numpy` to match version in `requirements.txt`?

**Review by jafermarq — COMMENTED** [review]
Hi @charlesbvll, some minor comments below. More concerning is the fact that I'm getting segmentation faults when running this on Ubuntu x64 (no matter if I use or don't use GPU) (but no issues on mac arm64). Seems to be related to the dataloaders...

-----
Update: setting `num_workers=0` in the dataloaders seems to fix the issue in Ubuntu. But this slows down training so I don't think it's an ideal solution

**Comment by charlesbvll** [comment]
> Hi @charlesbvll, some minor comments below. More concerning is the fact that I'm getting segmentation faults when running this on Ubuntu x64 (no matter if I use or don't use GPU) (but no issues on mac arm64). Seems to be related to the dataloaders...
> 
> Update: setting `num_workers=0` in the dataloaders seems to fix the issue in Ubuntu. But this slows down training so I don't think it's an ideal solution

@jafermarq I don't know of any other solutions to fix `num_workers > 0` related errors, even if the training is slower, I think it's better to set `num_workers=0` for stability.

File: examples/quickstart-monai/data.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        MedNISTDataset(trainX, trainY, train_transforms), num_shards, index
+    )
+    train_loader = DataLoader(train_ds, batch_size=300, shuffle=True)
+
+    val_ds = _partition(MedNISTDataset(valX, valY, val_transforms), num_shards, index)
+    val_loader = DataLoader(val_ds, batch_size=300)
+
+    test_ds = _partition(
+        MedNISTDataset(testX, testY, val_transforms), num_shards, index
+    )
```
Comment (jafermarq) [inline examples/quickstart-monai/data.py:54]:
```suggestion


def _partition(files_list, labels_list, num_shards, index):
    total_size = len(files_list)
    assert total_size == len(labels_list), f"List of datapoints and labels must be of the same length"
    shard_size = total_size // num_shards

    # Calculate start and end indices for the shard
    start_idx = index * shard_size
    if index == num_shards - 1:
        # Last shard takes the remainder
        end_idx = total_size
    else:
        end_idx = start_idx + shard_size

    # Create a subset for the shard
    files = files_list[start_idx: end_idx]
    labels = labels_list[start_idx: end_idx]
    return files, labels


def load_data(num_shards, index):
    image_file_list, image_label_list, num_total, num_class = _download_data()

    # get partition given index
    files_list, labels_list = _partition(image_file_list, image_label_list, num_shards, index)


    trainX, trainY, valX, valY, testX, testY = _split_data(
        files_list, labels_list, len(files_list)
    )
    train_transforms, val_transforms = _get_transforms()

    train_ds = MedNISTDataset(trainX, trainY, train_transforms)
    train_loader = DataLoader(train_ds, batch_size=300, shuffle=True)

    val_ds = MedNISTDataset(valX, valY, val_transforms)
    val_loader = DataLoader(val_ds, batch_size=300)

    test_ds = MedNISTDataset(testX, testY, val_transforms)
    test_loader = DataLoader(test_ds, batch_size=300)

    return train_loader, val_loader, test_loader, num_class
```

Comment (jafermarq) [reply examples/quickstart-monai/data.py:54]: While trying to rewrite how datasets are created (attempting to fix the segfault issue) I came up with this slightly simplified version of `load_data`. In short: partition first, then create datasets (as opposed of creating full datasets and then extract subsets)

**Review by jafermarq — COMMENTED** [review]
Let's keep the default `num_workers`

File: examples/quickstart-monai/README.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+python3 client.py --node-id 0
+```
+
+Start client 2 in the second terminal:
+
+```shell
+python3 client.py --node-id 1
+```
+
+You will see that the federated training is starting. Look at the [code](https://github.com/adap/flower/tree/main/examples/quickstart-monai) for a detailed explanation.
```
Comment (jafermarq) [inline examples/quickstart-monai/README.md:85]: I presume this is here so people reading this readme through the [Flower Example](https://flower.dev/docs/examples/) docs can go to the code directly? Would it make sense to add (like we did for the baselines) a button at the top of all `Flower Example` pages ? (in a different PR ob)

**Review by jafermarq — DISMISSED** [review]

**Comment by jafermarq** [comment]
@charlesbvll , worth adding a link to the top-level `README.md` to the example ?

File: examples/quickstart-monai/README.md (hunk 5 above)
Comment (charlesbvll) [reply examples/quickstart-monai/README.md:85]: Yes, you're right, this is a must have!

**Review by danieljanes — APPROVED** [review]

**Pull request merged**